File: restdoc-http-client/src/main/java/restdoc/client/invoke/HttpInvoker.java
package restdoc.client.invoke;

import org.springframework.core.env.Environment;
import org.springframework.http.HttpEntity;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;
import restdoc.remoting.common.body.HttpCommunicationCaptureBody;

import java.util.Map;
import java.util.Objects;

public class HttpInvoker {

    private final RestTemplate restTemplate;

    private final int port;

    private final String contextPath;

    public HttpInvoker(RestTemplate restTemplate, Environment environment) {
        this.restTemplate = restTemplate;
        port = Integer.parseInt(Objects.requireNonNull(environment.getProperty("server.port")));
        contextPath = environment.getProperty("server.servlet.context-path");
    }

    public ResponseEntity<Object> execute(HttpCommunicationCaptureBody capture) {
        String url = this.autocompleteURL(capture.getUrl());
        capture.setCompleteUrl(url);
        HttpEntity<Map<String, Object>> httpEntity =
                new HttpEntity<>(capture.getRequestBody(), capture.getRequestHeaders());

        return restTemplate.exchange(url, capture.getMethod(), httpEntity, Object.class, capture.getUriVariables());
    }


    private String autocompleteURL(String originURL) {
        if (originURL.startsWith("http") || originURL.startsWith("https")) return originURL;
        if (originURL.startsWith(contextPath)) {
            return String.format("http://127.0.0.1:%d%s", port, originURL);
        } else {
            if (originURL.startsWith("/")) {
                return String.format("http://127.0.0.1:%d%s%s", port, contextPath, originURL);
            } else {
                return String.format("http://127.0.0.1:%d%s/%s", port, contextPath, originURL);
            }
        }
    }
}

File: restdoc-rpc-client/restdoc-restweb-client/src/main/java/restdoc/client/restweb/invoke/HttpInvoker.kt
package restdoc.client.restweb.invoke

import org.springframework.core.env.Environment
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.ResponseEntity
import org.springframework.web.client.RestTemplate
import restdoc.remoting.common.body.HttpCommunicationCaptureBody
import java.util.*

open class HttpInvoker(private val restTemplate: RestTemplate, environment: Environment) {
    private val port: Int
    private val contextPath: String
    fun execute(capture: HttpCommunicationCaptureBody): ResponseEntity<Any> {
        val url = autocompleteURL(capture.url)
        capture.completeUrl = url
        val requestHeaders = HttpHeaders()
        capture.requestHeader.forEach { (key: String?, value: List<String?>?) -> requestHeaders[key] = value }
        val httpEntity = HttpEntity(capture.requestBody, requestHeaders)
        return restTemplate.exchange(url, capture.method, httpEntity, Any::class.java, capture.uriVariables)
    }

    private fun autocompleteURL(originURL: String): String {
        if (originURL.startsWith("http") || originURL.startsWith("https")) return originURL
        return if (originURL.startsWith(contextPath)) {
            String.format("http://127.0.0.1:%d%s", port, originURL)
        } else {
            if (originURL.startsWith("/")) {
                String.format("http://127.0.0.1:%d%s%s", port, contextPath, originURL)
            } else {
                String.format("http://127.0.0.1:%d%s/%s", port, contextPath, originURL)
            }
        }
    }

    init {
        port = Objects.requireNonNull(environment.getProperty("server.port")).toInt()
        contextPath = environment.getProperty("server.servlet.context-path")
    }
}